jsonrpc/throttling.go

```diff
@@ -1,70 +1,41 @@
 package jsonrpc
 
 import (
-	"container/heap"
+	"context"
 	"errors"
-	"sync"
 	"time"
+
+	"golang.org/x/sync/semaphore"
 )
 
 var errRequestLimitExceeded = errors.New("request limit exceeded")
 
+// Throttling provides functionality which limits number of concurrent requests
 type Throttling struct {
-	requestsPerSeconds int
-	requests           timeQueue
-	lock               sync.Mutex
+	sem     *semaphore.Weighted
+	timeout time.Duration
 }
 
-func NewThrottling(requestsPerSeconds int) *Throttling {
+// NewThrottling creates new throttling and limits number of concurrent requests to maximumConcurrentRequests
+func NewThrottling(maximumConcurrentRequests uint64, timeout time.Duration) *Throttling {
 	return &Throttling{
-		requestsPerSeconds: requestsPerSeconds,
+		sem:     semaphore.NewWeighted(int64(maximumConcurrentRequests)),
+		timeout: timeout,
 	}
 }
 
-func (t *Throttling) AttemptRequest() error {
-	t.lock.Lock()
-	defer t.lock.Unlock()
-
-	currTime := time.Now().UTC()
+// AttemptRequest returns an error if there is more thab
+func (t *Throttling) AttemptRequest(
+	parentCtx context.Context,
+	requestHandler func() (interface{}, error)) (interface{}, error) {
```

Comment:
Functional approach is handy when we don't know which functions are going to be throttled. In this situation, we implement every use case of `AttemptRequest()`, so the functional approach is not necessary.

Understood that the alternative is not that much better re verbosity because we would have to make the same call in each debug related function (including error checks).

Just a thought. Not suggesting changes. Both ways work 👍